crates/xline/src/server/command.rs

```diff
@@ -338,76 +338,71 @@ impl CommandExecutor {
 type AfterSyncResult = Result<AfterSyncOk<Command>, <Command as CurpCommand>::Error>;
 
 /// Collection of after sync results
-struct ASResultStates<'a> {
-    /// After sync cmds
-    cmds: Vec<AfterSyncCmd<'a, Command>>,
-    /// After sync results
-    results: Vec<Option<AfterSyncResult>>,
+struct ASResults<'a> {
+    /// After sync cmds and there execution results
+    cmd_results: Vec<(AfterSyncCmd<'a, Command>, Option<AfterSyncResult>)>,
 }
 
-impl<'a> ASResultStates<'a> {
+impl<'a> ASResults<'a> {
     /// Creates a new [`ASResultStates`].
     fn new(cmds: Vec<AfterSyncCmd<'a, Command>>) -> Self {
         Self {
-            results: iter::repeat_with(|| None::<AfterSyncResult>)
-                .take(cmds.len())
-                .collect(),
-            cmds,
+            // Initially all commands have no results
+            cmd_results: cmds.into_iter().map(|cmd| (cmd, None)).collect(),
         }
     }
 
+    #[allow(clippy::pattern_type_mismatch)] // can't be fixed
     /// Updates the results of commands that have errors by applying a given
     /// operation.
     fn update_err<F>(&mut self, op: F)
     where
         F: Fn(&AfterSyncCmd<'_, Command>) -> Result<(), ExecuteError>,
     {
-        for (cmd, result_opt) in self
-            .cmds
-            .iter()
-            .zip(self.results.iter_mut())
-            .filter(Self::filter_ok)
-        {
+        self.map_results(|(cmd, result_opt)| {
             if let Err(e) = op(cmd) {
                 let _ignore = result_opt.replace(Err(e));
             }
-        }
+        });
     }
 
     /// Updates the results of commands by applying a given operation.
+    #[allow(clippy::pattern_type_mismatch)] // can't be fixed
     fn update_result<F>(&mut self, op: F)
     where
         F: Fn(&AfterSyncCmd<'_, Command>) -> AfterSyncResult,
     {
-        for (cmd, result_opt) in self
-            .cmds
-            .iter()
-            .zip(self.results.iter_mut())
-            .filter(Self::filter_ok)
-        {
+        self.map_results(|(cmd, result_opt)| {
             let _ignore = result_opt.replace(op(cmd));
-        }
+        });
     }
 
-    /// Skip if the command execution has already errored
-    #[allow(clippy::pattern_type_mismatch)] // Can't be fixed
-    fn filter_ok(
-        (_cmd, result_opt): &(&AfterSyncCmd<'a, Command>, &mut Option<AfterSyncResult>),
-    ) -> bool {
-        result_opt.as_ref().is_none()
+    /// Applies a given operation to each command-result pair in `cmd_results` where the result is `None`.
+    #[allow(clippy::pattern_type_mismatch)] // can't be fixed
+    fn map_results<F>(&mut self, op: F)
```

Comment:
does `result == None` represents *waiting for after sync result* status? 
I think the name `map_results` is a bit confusing because it just apply op to pairs which has no result. And the `for_each` is also differ from `map`.

Reply:
Renamed to `for_each_none_result`.